File: src/binary-search/875/solution.py
import math

class Solution:
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        left, right = 1, max(piles)
        res = right
        
        
        while left <= right:
            k = (left + right) // 2
            hours_needed = sum(math.ceil(x / k) for x in piles)
            if hours_needed <= h:
                right = k - 1
                res = min(res, k)
            else:
                left = k + 1
        
        return res
                
            
        
s = Solution()
assert s.minEatingSpeed([3,6,7,11], 8) == 4
assert s.minEatingSpeed([30,11,23,4,20], 5) == 30
assert s.minEatingSpeed([30,11,23,4,20], 6) == 23

File: src/binary-search/875/solution.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <vector>

class Solution {
  public:
	int minEatingSpeed(std::vector<int>& piles, int h) {
		int l = 1;
		int r = *std::max_element(piles.begin(), piles.end());
		while (l < r) {
			int mid = l + (r - l) / 2;
			long long hours = 0;
			for (int p : piles) {
				hours += (long long)std::ceil((double)p / mid);
			}
			if (hours <= h) r = mid;
			else l = mid + 1;
		}
		return l;
	}
};

int main() {
	Solution s;

	std::vector<int> input1 = {3, 6, 7, 11};
	assert(s.minEatingSpeed(input1, 8) == 4);

	std::vector<int> input2 = {30, 11, 23, 4, 20};
	assert(s.minEatingSpeed(input2, 5) == 30);

	std::vector<int> input3 = {30, 11, 23, 4, 20};
	assert(s.minEatingSpeed(input3, 6) == 23);

	std::vector<int> input4 = {1};
	assert(s.minEatingSpeed(input4, 1) == 1);

	std::vector<int> input5 = {1000000000};
	assert(s.minEatingSpeed(input5, 2) == 500000000);

	std::vector<int> input6 = {312884470};
	assert(s.minEatingSpeed(input6, 968709470) == 1);

	std::vector<int> input7 = {1, 1, 1, 1};
	assert(s.minEatingSpeed(input7, 4) == 1);

	std::vector<int> input8 = {1, 1, 1, 1};
	assert(s.minEatingSpeed(input8, 5) == 1);

	std::vector<int> input9 = {5, 5, 5, 5};
	assert(s.minEatingSpeed(input9, 5) == 5);

	std::vector<int> input10 = {3, 6, 7, 11};
	assert(s.minEatingSpeed(input10, 4) == 11); // h == n → нужна max(piles)

	return 0;
}
